File: a2ui/datamodel.go
package a2ui

import (
	"fmt"
	"strconv"
	"strings"
)

// ParsePointer splits an A2UI data-binding path into its decoded reference
// tokens. It accepts standard RFC 6901 JSON Pointers (leading "/") as well as
// the relative paths A2UI permits inside collection scopes (no leading "/").
// The root pointer ("" or "/") yields a nil token slice.
func ParsePointer(path string) []string {
	if path == "" || path == "/" {
		return nil
	}
	trimmed := strings.TrimPrefix(path, "/")
	parts := strings.Split(trimmed, "/")
	for i, p := range parts {
		parts[i] = unescapeToken(p)
	}
	return parts
}

// unescapeToken decodes the RFC 6901 escape sequences ~1 -> "/" and ~0 -> "~".
func unescapeToken(t string) string {
	t = strings.ReplaceAll(t, "~1", "/")
	t = strings.ReplaceAll(t, "~0", "~")
	return t
}

// escapeToken encodes a reference token per RFC 6901.
func escapeToken(t string) string {
	t = strings.ReplaceAll(t, "~", "~0")
	t = strings.ReplaceAll(t, "/", "~1")
	return t
}

// JoinPointer builds an absolute JSON Pointer from decoded reference tokens.
func JoinPointer(tokens ...string) string {
	if len(tokens) == 0 {
		return "/"
	}
	var b strings.Builder
	for _, t := range tokens {
		b.WriteByte('/')
		b.WriteString(escapeToken(t))
	}
	return b.String()
}

// GetData resolves a data-binding path against a data model and returns the
// value found there. The second result reports whether the path resolved.
func GetData(model any, path string) (any, bool) {
	cur := model
	for _, tok := range ParsePointer(path) {
		switch node := cur.(type) {
		case map[string]any:
			v, ok := node[tok]
			if !ok {
				return nil, false
			}
			cur = v
		case []any:
			idx, _, err := arrayIndex(tok, len(node))
			if err != nil || idx < 0 || idx >= len(node) {
				return nil, false
			}
			cur = node[idx]
		default:
			return nil, false
		}
	}
	return cur, true
}

// ApplyUpdate applies an updateDataModel operation to a data model, returning
// the resulting model (which may be a brand-new root). When hasValue is false
// the key at path is removed. The empty/root path replaces (or clears) the
// entire model. Intermediate map nodes are created as needed.
func ApplyUpdate(model any, path string, value any, hasValue bool) (any, error) {
	tokens := ParsePointer(path)
	if len(tokens) == 0 {
		if !hasValue {
			return nil, nil
		}
		return value, nil
	}
	if model == nil {
		model = map[string]any{}
	}
	return applyTokens(model, tokens, value, hasValue)
}

func applyTokens(node any, tokens []string, value any, hasValue bool) (any, error) {
	key := tokens[0]
	last := len(tokens) == 1

	switch n := node.(type) {
	case map[string]any:
		if last {
			if hasValue {
				n[key] = value
			} else {
				delete(n, key)
			}
			return n, nil
		}
		child, ok := n[key]
		if !ok || child == nil {
			child = map[string]any{}
		}
		updated, err := applyTokens(child, tokens[1:], value, hasValue)
		if err != nil {
			return nil, err
		}
		n[key] = updated
		return n, nil

	case []any:
		idx, isAppend, err := arrayIndex(key, len(n))
		if err != nil {
			return nil, err
		}
		if last {
			if isAppend {
				if hasValue {
					n = append(n, value)
				}
				return n, nil
			}
			if idx < 0 || idx >= len(n) {
				return nil, fmt.Errorf("a2ui: array index %q out of range", key)
			}
			if hasValue {
				n[idx] = value
			} else {
				// Per spec: removing an array element sets it to undefined,
				// preserving length.
				n[idx] = nil
			}
			return n, nil
		}
		if isAppend || idx < 0 || idx >= len(n) {
			return nil, fmt.Errorf("a2ui: cannot descend into array index %q", key)
		}
		child := n[idx]
		if child == nil {
			child = map[string]any{}
		}
		updated, err := applyTokens(child, tokens[1:], value, hasValue)
		if err != nil {
			return nil, err
		}
		n[idx] = updated
		return n, nil

	default:
		// Scalar (or nil) where a container is required: replace with a map so
		// the path can be created.
		return applyTokens(map[string]any{}, tokens, value, hasValue)
	}
}

// arrayIndex parses an array reference token. The special token "-" addresses
// the position one past the end (append). It returns the index, whether the
// token was the append marker, and any parse error.
func arrayIndex(tok string, length int) (idx int, isAppend bool, err error) {
	if tok == "-" {
		return length, true, nil
	}
	i, err := strconv.Atoi(tok)
	if err != nil {
		return 0, false, fmt.Errorf("a2ui: invalid array index %q", tok)
	}
	return i, false, nil
}

File: desktop/frontend/src/a2ui-renderer/datamodel.ts
/**
 * JSON Pointer (RFC 6901) 数据模型引擎。
 *
 * 对齐 Go 端 a2ui/datamodel.go 的语义：
 *   - ParsePointer → parsePointer
 *   - GetData → getData
 *   - ApplyUpdate → applyUpdate
 *   - arrayIndex → arrayIndex
 */

// ── 指针解析 ───────────────────────────────────────

const _unescapeToken = (t: string): string =>
  t.replace(/~1/g, '/').replace(/~0/g, '~')

const _escapeToken = (t: string): string =>
  t.replace(/~/g, '~0').replace(/\//g, '~1')

/**
 * 将 JSON Pointer 拆分为解码后的 token 数组。
 * 根指针 "" 或 "/" 返回空数组。
 */
export function parsePointer(path: string): string[] {
  if (path === '' || path === '/') return []
  const trimmed = path.startsWith('/') ? path.slice(1) : path
  return trimmed.split('/').map(_unescapeToken)
}

/**
 * 从解码后的 token 数组构建绝对 JSON Pointer。
 */
export function joinPointer(...tokens: string[]): string {
  if (tokens.length === 0) return '/'
  return '/' + tokens.map(_escapeToken).join('/')
}

// ── 数据读取 ───────────────────────────────────────

/**
 * 按 JSON Pointer 路径从数据模型中取值。
 * 返回 [value, found]。
 */
export function getData(model: unknown, path: string): [unknown, boolean] {
  let cur: unknown = model
  for (const tok of parsePointer(path)) {
    if (cur === null || typeof cur !== 'object') return [undefined, false]
    if (Array.isArray(cur)) {
      const [idx] = _arrayIndex(tok, cur.length)
      if (idx < 0 || idx >= cur.length) return [undefined, false]
      cur = cur[idx]
    } else {
      const m = cur as Record<string, unknown>
      if (!(tok in m)) return [undefined, false]
      cur = m[tok]
    }
  }
  return [cur, true]
}

// ── 数据更新 ───────────────────────────────────────

/**
 * 在数据模型上应用 updateDataModel 操作。
 * hasValue=true 时设值，false 时删除 key（数组元素置为 undefined）。
 * 返回新的数据模型根对象（不可变更新）。
 */
export function applyUpdate(
  model: unknown,
  path: string,
  value: unknown,
  hasValue: boolean,
): unknown {
  const tokens = parsePointer(path)
  if (tokens.length === 0) {
    return hasValue ? value : null
  }
  const root = model ?? ({} as Record<string, unknown>)
  return _applyTokens(structuredClone(root), tokens, value, hasValue)
}

function _applyTokens(
  node: unknown,
  tokens: string[],
  value: unknown,
  hasValue: boolean,
): unknown {
  const key = tokens[0]
  const last = tokens.length === 1

  if (Array.isArray(node)) {
    const [idx, isAppend] = _arrayIndex(key, node.length)
    if (last) {
      if (isAppend) {
        if (hasValue) node.push(value)
        return node
      }
      if (idx < 0 || idx >= node.length) return node
      if (hasValue) {
        node[idx] = value
      } else {
        node[idx] = undefined
      }
      return node
    }
    if (isAppend || idx < 0 || idx >= node.length) return node
    const child = node[idx] ?? ({} as Record<string, unknown>)
    node[idx] = _applyTokens(child, tokens.slice(1), value, hasValue)
    return node
  }

  const m = node as Record<string, unknown>
  if (last) {
    if (hasValue) {
      m[key] = value
    } else {
      delete m[key]
    }
    return m
  }
  const child = m[key] ?? ({} as Record<string, unknown>)
  m[key] = _applyTokens(child, tokens.slice(1), value, hasValue)
  return m
}

// ── 辅助 ───────────────────────────────────────────

/**
 * 解析数组索引 token。"-" 表示 append。
 * 返回 [index, isAppend]。
 */
function _arrayIndex(tok: string, length: number): [number, boolean] {
  if (tok === '-') return [length, true]
  const idx = parseInt(tok, 10)
  if (isNaN(idx)) return [-1, false]
  return [idx, false]
}
